FWCore/src/components/PileupMerge.cpp

```diff
@@ -0,0 +1,117 @@
+#include "PileupMerge.h"
+
+
+DECLARE_ALGORITHM_FACTORY(PileupMerge)
+
+PileupMerge::PileupMerge(const std::string& aName, ISvcLocator* aSvcLoc):
+  GaudiAlgorithm(aName, aSvcLoc),
+  m_store(),
+  m_reader() {
+  declareProperty("pileupGenParticles", m_genParticlesOut, "Generated particles collection (output)");
+  declareProperty("pileupGenVertices", m_genVerticesOut, "Generated vertices collection (output)");
+  declareProperty("pileupTrackHits", m_trackHitsOut, "Track Hit Collection (output)");
+  declareProperty("pileupPosTrackHits", m_posTrackHits, "Positioned Track Hits Collection (output)");
+  declareProperty("pileupCaloHits", m_caloHitsOut, "Calo Hit Collection (output)");
+  declareProperty("pileupPosCaloHits", m_posCaloHits, "Positioned Calo Hits Collection (output)");
+  declareProperty("PileUpTool", m_pileUpTool);
+}
+
+StatusCode PileupMerge::initialize() {
+  currentEvent = 0;
+  StatusCode sc = GaudiAlgorithm::initialize();
+  m_reader.openFile(m_pileupFilename);
+  m_store.setReader(&m_reader);
+  IRndmGenSvc* randSvc = svc<IRndmGenSvc>("RndmGenSvc", true);
+  sc = m_flatDist.initialize(randSvc, Rndm::Flat(0., 1.));
+  return sc;
+}
+
+
+StatusCode PileupMerge::finalize() {
+  StatusCode sc = GaudiAlgorithm::finalize();
+  return sc;
+}
+
+
+StatusCode PileupMerge::execute() {
+  unsigned nEvents = m_reader.getEntries();
+
+
+  const fcc::TrackHitCollection* trackHitCollection;
+  fcc::TrackHitCollection* hitCollOut = new fcc::TrackHitCollection();
+
+  const fcc::PositionedTrackHitCollection* posTrackHitCollection;
+  fcc::PositionedTrackHitCollection* posHitCollOut = new fcc::PositionedTrackHitCollection();
+
+  const fcc::MCParticleCollection* mcParticleCollection;
+  fcc::MCParticleCollection* mcPartCollOut = new fcc::MCParticleCollection();
+
+  const fcc::GenVertexCollection* genVertexCollection;
+  fcc::GenVertexCollection* genVertCollOut = new fcc::GenVertexCollection();
+
+  const fcc::CaloHitCollection* caloHitCollection;
+  fcc::CaloHitCollection* caloHitCollOut = new fcc::CaloHitCollection();
+
+  const fcc::PositionedCaloHitCollection* posCaloHitCollection;
+  fcc::PositionedCaloHitCollection* posCaloHitCollOut = new fcc::PositionedCaloHitCollection();
+
+  const unsigned int numPileUp = m_pileUpTool->numberOfPileUp();
+  for(unsigned iev = 0; iev < numPileup; ++iev) {
+
+    if ( m_randomizePileup == true ) {
+      if (m_flatDist() < 0.3) { // skip one in three events
+        ++currentEvent;
+      }
+    }
+    currentEvent = (currentEvent + 1) % nEvents; // start over from beginning if necessary
```

Comment:
Hmmm, I think the behavior is correct, but the variable names / documentation are a bit misleading here. For every signal event (call to `execute()`) this code is called `numPileUp` times. `currentEvent` should probably be called `minBiasEventPoolIndex` as it determines which event is read next from the min. bias event pool. I'll update that!